README.MD
# weweChat

<img src="https://github.com/trazyn/weweChat/blob/master/resource/128x128.png" />

> Icon by [octiviotti](https://octiviotti.deviantart.com/)

[![Build Status](https://travis-ci.org/trazyn/weweChat.svg?branch=master)](https://travis-ci.org/trazyn/weweChat)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com)
[![PRs Welcome](https://img.shields.io/badge/PRs-Welcome-brightgreen.svg)](https://github.com/trazyn/weweChat/pulls)

Unofficial WeChat client built with React, MobX and Electron.

API from [https://web.wechat.com/](https://web.wechat.com/)

## [CHANGELOG](https://github.com/trazyn/weweChat/blob/master/CHANGELOG.MD)

## Feature
- Work on desktop
- On macOS, window vibrancy effect
- Block message recall(Default settings is not block)
- Desktop notifications
- Keyboard shortcuts supported
- Send image by paste
  ![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/pasteconfirmation.png)
- Drag to send file
  ![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/dragdrop.png)
- Batch send message
  ![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/batchsend.png)
- Send GIF emoji
  ![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/sendgif.gif)

## Install
Download the last version on the [website](https://github.com/trazyn/weweChat/releases/latest) or below.

#### Mac(10.9+)
[Download](https://github.com/trazyn/weweChat/releases/download/1.0.0/wewechat-1.0.0-mac.dmg) the `.dmg` file.
Or use [Homebrew-Cask](https://caskroom.github.io/):
```
$ brew cask install wewechat
```

#### Linux
[Download](https://github.com/trazyn/weweChat/releases/download/1.0.0/wewechat-1.0.0-x86_64.AppImage) the `.AppImage` file.
The AppImage needs to be made executable after download:
```
$ chmod u+x wewechat-1.0.0-x86_64.AppImage
```

#### Windows
[Download](https://github.com/trazyn/weweChat/releases/download/1.0.0/wewechat-1.0.0-win-setup.exe) the `.exe` file.

## Screenshot
![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/0.png)
![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/1.png)
![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/2.png)
![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/3.png)
![preview](https://raw.githubusercontent.com/trazyn/weweChat/master/screenshots/4.png)

## Development
```
$ npm install
$ npm run dev
```

Generate the binary:
* For Linux
```
$ npm run package-linux
```
* For Mac
```
$ npm run package-mac
```
After that, you will see the binary in `./release` folder

## Keyboard shortcuts

Description            | Keys
-----------------------| -----------------------
New conversation       | <kbd>Cmd</kbd> <kbd>N</kbd>
Search conversations   | <kbd>Cmd</kbd> <kbd>F</kbd>
Hide conversation      | <kbd>Shift</kbd> <kbd>Cmd</kbd> <kbd>M</kbd>
Jump to conversation   | <kbd>Cmd</kbd> <kbd>0</kbd> ... <kbd>9</kbd>
Next conversation      | <kbd>Cmd</kbd> <kbd>J</kbd>
Previous conversation  | <kbd>Cmd</kbd> <kbd>K</kbd>
Batch message          | <kbd>Cmd</kbd> <kbd>B</kbd>
Toggle Full Screen     | <kbd>Shift</kbd> <kbd>Cmd</kbd> <kbd>F</kbd>
Insert QQ emoji        | <kbd>Cmd</kbd> <kbd>I</kbd>
Preferences            | <kbd>Cmd</kbd> <kbd>,</kbd>

### TODO
- [x] Windows support
- [x] Linux support
- [x] Sticky on top
- [x] Delete chat session
- [x] Mark as Read
- [x] Chat Room
  - [x] Show correct contact
  - [x] Show members
  - [x] Add / Remove member
  - [x] Create chat room
- [x] Receive message
  - [x] Text
  - [x] Image
  - [x] Voice
  - [x] Location
  - [x] Sticker
  - [x] Contact Card
  - [x] Video
  - [x] Money Transger
  - [x] Location sharing
  - [x] Download File
- [x] Send message
  - [x] Text
  - [x] File
  - [x] Image
  - [x] Video
  - [x] Recall
- [x] Forward text message
- [x] Forward emoji
- [x] Forward image
- [x] Forward file
- [x] Forward video
- [x] Show QQ emoji
- [x] Search chat set
- [x] Search and create chat room
- [x] Desktop notification
- [x] Keep online
- [x] Logout
- [x] Autosart at login
- [ ] Theme
- [x] Send image from clipboard
- [x] Drag to send file
- [x] Batch send message
- [ ] Mention a user
- [ ] Chat bot

### FAQ
- `Mac` 上如何修改图标，请参考 #39（另外问下，有没屌大的或者胸大的帮忙设计一个图标啊）
- 关于历史记录的问题请参考 #30


### License
MIT License



CHANGELOG.MD


# 1.1.4

### 2017-09-22

#### Improvements
- 发送消息后滚动到对话窗口最下面

#### Bug fixed
- 修复 #34 快速`enter`时会重复发送消息
- 修复点击状态栏小图标展开菜单后，收到消息后重新创建小图标导致程序退出的问题

# 1.1.3

### 2017-09-19

#### Bug fixed
- #36 #34 优化内存使用，修复由于1.1.1显示头像到菜单栏，接收新消息时刷新菜单中会话列表，重复创建nativeImage导致GC无法及时回收所造成内存占用过高


# 1.1.2

### 2017-09-18

#### Improvements
- 增加自动更新


# 1.1.1

### 2017-09-14

#### Improvements
- 增加用户头像到菜单栏

#### Bug fixed
- 修复登出时网络错误，导致未提示客户端已离线

### 2017-09-12
#29 替换`html`为对应`emoji`

### 2017-09-11
#28 在 windows 和 linux 上不显示菜单

### 2017-09-10

#### Bug fixed
- 联系人列表只显示好友，不显示临时联系人

### 2017-09-08

#### Improvements
- 添加文件助手, #27

#### Bug fixed
- 修改样式，防止文字信息内容过长

### 2017-09-07

#### Improvements
- 调整表情框表情大小
- 聊天窗口用户名取消换行

#### Feature
- 多文件拖拽发送

#### Bug fixed
- 仅在聊天开启时允许文件拖拽
- 修复手机端消息同步客户端错误
- 修复图片加载失败使用错误`src`进行回滚导致客户端卡死的问题


src/js/utils/emoji.js

const EmojiList = ['笑脸', '生病', '破涕为笑', '吐舌', '脸红', '恐惧', '失望', '无语', '嘿哈', '捂脸', '奸笑', '机智', '皱眉', '耶', '鬼魂', '合十', '强壮', '庆祝', '礼物', '红包', '鸡', '开心', '大笑', '热情', '眨眼', '色', '接吻', '亲吻', '露齿笑', '满意', '戏弄', '得意', '汗', '低落', '呸', '焦虑', '担心', '震惊', '悔恨', '眼泪', '哭', '晕', '心烦', '生气', '睡觉', '恶魔', '外星人', '心', '心碎', '丘比特', '闪烁', '星星', '叹号', '问号', '睡着', '水滴', '音乐', '火', '便便', '弱', '拳头', '胜利', '上', '下', '右', '左', '第一', '吻', '热恋', '男孩', '女孩', '女士', '男士', '天使', '骷髅', '红唇', '太阳', '下雨', '多云', '雪人', '月亮', '闪电', '海浪', '猫', '小狗', '老鼠', '仓鼠', '兔子', '狗', '青蛙', '老虎', '考拉', '熊', '猪', '牛', '野猪', '猴子', '马', '蛇', '鸽子', '鸡', '企鹅', '毛虫', '章鱼', '鱼', '鲸鱼', '海豚', '玫瑰', '花', '棕榈树', '仙人掌', '礼盒', '南瓜灯', '圣诞老人', '圣诞树', '礼物', '铃', '气球', 'CD', '相机', '录像机', '电脑', '电视', '电话', '解锁', '锁', '钥匙', '成交', '灯泡', '邮箱', '浴缸', '钱', '炸弹', '手枪', '药丸', '橄榄球', '篮球', '足球', '棒球', '高尔夫', '奖杯', '入侵者', '唱歌', '吉他', '比基尼', '皇冠', '雨伞', '手提包', '口红', '戒指', '钻石', '咖啡', '啤酒', '干杯', '鸡尾酒', '汉堡', '薯条', '意面', '寿司', '面条', '煎蛋', '冰激凌', '蛋糕', '苹果', '飞机', '火箭', '自行车', '高铁', '警告', '旗', '男人', '女人', 'O', 'X', '版权', '注册商标', '商标'];
const QQFaceMap = {
    '微笑': '0',
    '撇嘴': '1',
    '色': '2',
    '发呆': '3',
    '得意': '4',
    '流泪': '5',
    '害羞': '6',
    '闭嘴': '7',
    '睡': '8',
    '大哭': '9',
    '尴尬': '10',
    '发怒': '11',
    '调皮': '12',
    '呲牙': '13',
    '惊讶': '14',
    '难过': '15',
    '酷': '16',
    '冷汗': '17',
    '抓狂': '18',
    '吐': '19',
    '偷笑': '20',
    '可爱': '21',
    '愉快': '21',
    '白眼': '22',
    '傲慢': '23',
    '饥饿': '24',
    '困': '25',
    '惊恐': '26',
    '流汗': '27',
    '憨笑': '28',
    '悠闲': '29',
    '大兵': '29',
    '奋斗': '30',
    '咒骂': '31',
    '疑问': '32',
    '嘘': '33',
    '晕': '34',
    '疯了': '35',
    '折磨': '35',
    '衰': '36',
    '骷髅': '37',
    '敲打': '38',
    '再见': '39',
    '擦汗': '40',
    '抠鼻': '41',
    '鼓掌': '42',
    '糗大了': '43',
    '坏笑': '44',
    '左哼哼': '45',
    '右哼哼': '46',
    '哈欠': '47',
    '鄙视': '48',
    '委屈': '49',
    '快哭了': '50',
    '阴险': '51',
    '亲亲': '52',
    '吓': '53',
    '可怜': '54',
    '菜刀': '55',
    '西瓜': '56',
    '啤酒': '57',
    '篮球': '58',
    '乒乓': '59',
    '咖啡': '60',
    '饭': '61',
    '猪头': '62',
    '玫瑰': '63',
    '凋谢': '64',
    '嘴唇': '65',
    '示爱': '65',
    '爱心': '66',
    '心碎': '67',
    '蛋糕': '68',
    '闪电': '69',
    '炸弹': '70',
    '刀': '71',
    '足球': '72',
    '瓢虫': '73',
    '便便': '74',
    '月亮': '75',
    '太阳': '76',
    '礼物': '77',
    '拥抱': '78',
    '强': '79',
    '弱': '80',
    '握手': '81',
    '胜利': '82',
    '抱拳': '83',
    '勾引': '84',
    '拳头': '85',
    '差劲': '86',
    '爱你': '87',
    NO: '88',
    OK: '89',
    '爱情': '90',
    '飞吻': '91',
    '跳跳': '92',
    '发抖': '93',
    '怄火': '94',
    '转圈': '95',
    '磕头': '96',
    '回头': '97',
    '跳绳': '98',
    '投降': '99',
    '激动': '100',
    '乱舞': '101',
    '献吻': '102',
    '左太极': '103',
    '右太极': '104',
    '嘿哈': '105',
    '捂脸': '106',
    '奸笑': '107',
    '机智': '108',
    '皱眉': '109',
    '耶': '110',
    '鸡': '111',
    '红包': '112',
    Smile: '0',
    Grimace: '1',
    Drool: '2',
    Scowl: '3',
    Chill: '4',
    CoolGuy: '4',
    Sob: '5',
    Shy: '6',
    Shutup: '7',
    Silent: '7',
    Sleep: '8',
    Cry: '9',
    Awkward: '10',
    Pout: '11',
    Angry: '11',
    Wink: '12',
    Tongue: '12',
    Grin: '13',
    Surprised: '14',
    Surprise: '14',
    Frown: '15',
    Cool: '16',
    Ruthless: '16',
    Tension: '17',
    Blush: '17',
    Scream: '18',
    Crazy: '18',
    Puke: '19',
    Chuckle: '20',
    Joyful: '21',
    Slight: '22',
    Smug: '23',
    Hungry: '24',
    Drowsy: '25',
    Panic: '26',
    Sweat: '27',
    Laugh: '28',
    Loafer: '29',
    Commando: '29',
    Strive: '30',
    Determined: '30',
    Scold: '31',
    Doubt: '32',
    Shocked: '32',
    Shhh: '33',
    Dizzy: '34',
    Tormented: '35',
    BadLuck: '36',
    Toasted: '36',
    Skull: '37',
    Hammer: '38',
    Wave: '39',
    Relief: '40',
    Speechless: '40',
    DigNose: '41',
    NosePick: '41',
    Clap: '42',
    Shame: '43',
    Trick: '44',
    'Bah！L': '45',
    'Bah！R': '46',
    Yawn: '47',
    Lookdown: '48',
    'Pooh-pooh': '48',
    Wronged: '49',
    Shrunken: '49',
    Puling: '50',
    TearingUp: '50',
    Sly: '51',
    Kiss: '52',
    'Uh-oh': '53',
    Wrath: '53',
    Whimper: '54',
    Cleaver: '55',
    Melon: '56',
    Watermelon: '56',
    Beer: '57',
    Basketball: '58',
    PingPong: '59',
    Coffee: '60',
    Rice: '61',
    Pig: '62',
    Rose: '63',
    Wilt: '64',
    Lip: '65',
    Lips: '65',
    Heart: '66',
    BrokenHeart: '67',
    Cake: '68',
    Lightning: '69',
    Bomb: '70',
    Dagger: '71',
    Soccer: '72',
    Ladybug: '73',
    Poop: '74',
    Moon: '75',
    Sun: '76',
    Gift: '77',
    Hug: '78',
    Strong: '79',
    ThumbsUp: '79',
    Weak: '80',
    ThumbsDown: '80',
    Shake: '81',
    Victory: '82',
    Peace: '82',
    Admire: '83',
    Fight: '83',
    Salute: '83',
    Beckon: '84',
    Fist: '85',
    Pinky: '86',
    Love: '2',
    RockOn: '87',
    No: '88',
    'Nuh-uh': '88',
    InLove: '90',
    Blowkiss: '91',
    Waddle: '92',
    Tremble: '93',
    'Aaagh!': '94',
    Twirl: '95',
    Kotow: '96',
    Lookback: '97',
    Dramatic: '97',
    Jump: '98',
    JumpRope: '98',
    'Give-in': '99',
    Surrender: '99',
    Hooray: '100',
    HeyHey: '101',
    Meditate: '101',
    Smooch: '102',
    'TaiJi L': '103',
    'TaiChi L': '103',
    'TaiJi R': '104',
    'TaiChi R': '104',
    Hey: '105',
    Facepalm: '106',
    Smirk: '107',
    Smart: '108',
    Concerned: '109',
    'Yeah!': '110',
    Chicken: '111',
    Packet: '112',
    '發呆': '3',
    '流淚': '5',
    '閉嘴': '7',
    '尷尬': '10',
    '發怒': '11',
    '調皮': '12',
    '驚訝': '14',
    '難過': '15',
    '饑餓': '24',
    '累': '25',
    '驚恐': '26',
    '悠閑': '29',
    '奮鬥': '30',
    '咒罵': '31',
    '疑問': '32',
    '噓': '33',
    '暈': '34',
    '瘋了': '35',
    '骷髏頭': '37',
    '再見': '39',
    '摳鼻': '41',
    '羞辱': '43',
    '壞笑': '44',
    '鄙視': '48',
    '陰險': '51',
    '親親': '52',
    '嚇': '53',
    '可憐': '54',
    '籃球': '58',
    '飯': '61',
    '豬頭': '62',
    '枯萎': '64',
    '愛心': '66',
    '閃電': '69',
    '炸彈': '70',
    '甲蟲': '73',
    '太陽': '76',
    '禮物': '77',
    '擁抱': '78',
    '強': '79',
    '勝利': '82',
    '拳頭': '85',
    '差勁': '86',
    '愛你': '88',
    '愛情': '90',
    '飛吻': '91',
    '發抖': '93',
    '噴火': '94',
    '轉圈': '95',
    '磕頭': '96',
    '回頭': '97',
    '跳繩': '98',
    '激動': '100',
    '亂舞': '101',
    '獻吻': '102',
    '左太極': '103',
    '右太極': '104',
    '吼嘿': '105',
    '掩面': '106',
    '機智': '108',
    '皺眉': '109',
    '歐耶': '110',
    '雞': '111',
    '紅包': '112',
};

const emoji = [];

Object.keys(QQFaceMap).slice(0, 105).map(e => {
    var id = QQFaceMap[e];

    emoji.push({
        key: e,
        value: QQFaceMap[e],
        className: `qqemoji qqemoji${id}`
    });
});

function getEmojiClassName(name) {
    name = name.substring(1, name.length - 1);

    var keys = Object.keys(QQFaceMap);
    var key = keys.find(e => e === name);

    if (key) {
        let id = QQFaceMap[key];
        let emojiCN = keys.find(e => QQFaceMap[e] === id);

        if (+id < 200) {
            return `qqemoji qqemoji${id}`;
        }

        if (EmojiList.includes(emojiCN)) {
            return `emoji emoji${id}`;
        }
    }

    return false;
}

function parser(text) {
    var decodeText = text;

    (text.match(/\[[\w\s\u4E00-\u9FCC\u3400-\u4DB5\uFA0E\uFA0F\uFA11\uFA13\uFA14\uFA1F\uFA21\uFA23\uFA24\uFA27-\uFA29\ud840-\ud868\udc00-\udfff\ud869[\udc00-\uded6\udf00-\udfff\ud86a-\ud86c\udc00-\udfff\ud86d[\udc00-\udf34\udf40-\udfff\ud86e\udc00-\udc1d]+\]/g) || []).map(e => {
        var className = getEmojiClassName(e);

        if (!className) {
            // Invalid emoji
            return;
        }
        text = decodeText = text.split(`${e}`).join(`<a class="${className}"></a>`);
    });

    return normalize(decodeText);
}

function normalize(text = '') {
    var matchs = text.match(/<span class="emoji emoji[0-9a-fA-F]+"><\/span>/g) || [];
    var decodeText = text;

    try {
        matchs.map(e => {
            // Decode utf16 to emoji
            var emojiCode = e.match(/emoji([0-9a-fA-F]+)/)[1].substr(0, 5);
            var emoji = String.fromCodePoint(parseInt(emojiCode, 16));
            text = decodeText = text.split(e).join(emoji);
        });
    } catch (ex) {
        console.error('Failed decode %s: %o', text, ex);
    }

    return decodeText;
}

export { emoji, parser, normalize };
